### PR TITLE
Add Parallel::find and Parallel::find_result

### PR DESCRIPTION
**Pull request by ellcs**
Implementation of ticket #273.

### CODE REVIEW

File: lib/parallel.rb [hunk 1]
```diff
@@ -271,14 +297,22 @@ def map(source, options = {}, &block)
       add_progress_bar!(job_factory, options)
 
       results = if size == 0
-        work_direct(job_factory, options, &block)
-      elsif method == :in_threads
-        work_in_threads(job_factory, options.merge(:count => size), &block)
-      else
-        work_in_processes(job_factory, options.merge(:count => size), &block)
-      end
+                  work_direct(job_factory, options, &block)
```
Comment (grosser) [inline lib/parallel.rb:300]: 2-space plz

File: lib/parallel.rb [hunk 2]
```diff
@@ -485,7 +524,8 @@ def process_incoming_jobs(read, write, job_factory, options, &block)
     end
 
     def handle_exception(exception, results)
-      return nil if [Parallel::Break, Parallel::Kill].include? exception.class
+      return if [Parallel::Break, Parallel::Kill].include?(exception.class)
+      return exception if exception.class == Parallel::Return
```
Comment (grosser) [inline lib/parallel.rb:528]: I'd expect this to return the value ?
```suggestion
      return exception.result if exception.class == Parallel::Return
```

File: lib/parallel.rb [hunk 3]
```diff
@@ -271,14 +297,22 @@ def map(source, options = {}, &block)
...
+        if options[:return_one] 
+          if results.is_a?(Return)
+            results.return_value
+          end
+        elsif options[:return_results] 
+          results
+        else
+          source
+        end
       end
```
Comment (grosser) [inline lib/parallel.rb:316]: this looks like it's doing similar kind of logic to `handle_exception` ... can it be combined ?

File: lib/parallel.rb [hunk 4]
```diff
@@ -228,6 +236,24 @@ def each(array, options={}, &block)
...
+
+    def find(array, options = {}, &block)
+      raise "You must provide a block when calling #find" if block.nil?
+      map(array, options.merge({return_one: true})) do |*a| 
+        if block.call(*a) 
+          raise Parallel::Return.new(*a)
+        end
+      end
+    end
+
```
Comment (grosser) [inline lib/parallel.rb:256]: maybe let's just do the `raise Return` part first, so this is a smaller PR ?

File: lib/parallel.rb [hunk 5]
```diff
@@ -14,6 +14,14 @@ class Break < StandardError
   class Kill < StandardError
   end
 
+  class Return < StandardError
+    attr_reader :return_value
```
Comment (grosser) [inline lib/parallel.rb:18]:
```suggestion
    attr_reader :value
```

**Comment by grosser** [comment]
Logic looks right, but I think you stumbled into a problem area of the code there, I think we should do a cleanup first before adding a new feature, the handle_exception and results/source logic seems like they would go together

**Comment by ellcs** [comment]
Thanks for the review. I will take the time to apply some changes soon.

**Comment by grosser** [comment]
v1.20 has `Parallel.map([1, 2, 3]) { |i| raise Parallel::Break, i if i == 2 } == 2`

**Pull request closed without merging**